### ISSUE TITLE
list[int] incorrectly rejected as argument to parameter typed list[str | int]

### ISSUE THREAD

**Issue by vitalii-diachkov-sumup** [astral-sh/ty#2856]
list[int] should be accepted where list[str | int] is expected, but ty reports an invalid-argument-type error.

This suggests incorrect handling of union types inside generic type arguments.

Here is [`ty` playground URL](https://play.ty.dev/4d651115-4dc2-454d-8fc6-9829f49b8d5d)

I would expect `list[int]` to be allowed in this context as it can be used anywhere `list[str | int]` can be used

**Comment by AlexWaygood** [comment]
Hi! ty is working as expected here, because `list` is invariant. Every other type checker reports the same error:

```
% uvx mypy foo.py   
foo.py:5: error: Argument 1 to "foo" has incompatible type "list[int]"; expected "list[str | int]"  [arg-type]
foo.py:5: note: "list" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
foo.py:5: note: Consider using "Sequence" instead, which is covariant
Found 1 error in 1 file (checked 1 source file)

% uvx pyright foo.py
/Users/alexw/dev/ruff/foo.py
  /Users/alexw/dev/ruff/foo.py:5:5 - error: Argument of type "list[int]" cannot be assigned to parameter "l" of type "list[str | int]" in function "foo"
    "list[int]" is not assignable to "list[str | int]"
      Type parameter "_T@list" is invariant, but "int" is not the same as "str | int"
      Consider switching from "list" to "Sequence" which is covariant (reportArgumentType)
1 error, 0 warnings, 0 informations

% uvx pyrefly check foo.py
Installed 1 package in 4ms
ERROR Argument `list[int]` is not assignable to parameter `l` with type `list[int | str]` in function `foo` [bad-argument-type]
 --> foo.py:5:5
  |
5 | foo(arg)
  |     ^^^
  |
 INFO 1 error

% cat foo.py
def foo(l: list[str | int]) -> int:
    return len(l)

arg: list[int] = [1, 2, 3]
foo(arg)
```

We don't have great docs on variance yet, unfortunately, but the mypy docs are pretty strong on this topic:
- https://mypy.readthedocs.io/en/stable/common_issues.html#variance
- https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generics

**Comment by vitalii-diachkov-sumup** [comment]
@AlexWaygood I am asking because I was now migrating my project from `mypy` and I had a call to qdrant's `delete(...)` method of `QdrantClient`. It accepts `points_selector=...` argument of `PointsSelector` type that is a union with one of the types being `PointIdsList`, and I had:

```python
    def _delete(self, ids: list[int]) -> None:
        await self._client.delete(
            collection_name=self._collection_name,
            points_selector=models.PointIdsList(points=ids),
            wait=True,
        )
```

With `mypy` I never had an error, with `ty` it is saying that `list[int]` (type of `ids` argument here) is not valid becuase `points` expects `list[str | int]` at the end and I don't understand why `list[int]` cannot be a valid type in this case 🤔

Here's the type from qdrant client lib:
https://github.com/qdrant/qdrant-client/blob/e7101dc5baf4d3ce7d6321e1500311619aa95a99/qdrant_client/http/models/models.py#L2169

**Comment by AlexWaygood** [comment]
> With `mypy` I never had an error, with `ty` it is saying that `list[int]` (type of `ids` argument here) is not valid becuase `points` expects `list[str | int]` at the end and I don't understand why `list[int]` cannot be a valid type in this case 🤔

There must be another factor in `qdrant-client`'s types that explains the difference between mypy and ty here. For the minimal repro you posted in your original comment, mypy/pyright/pyrefly/ty all behave the same, and the documentation links I posted for you above explain why this behaviour is correct.

If you're able to find a minimal repro that demonstrates a difference in behaviour between mypy and ty that you think demonstrates a bug in ty, please open a new issue :-)

**Comment by vitalii-diachkov-sumup** [comment]
@AlexWaygood I have actually tried to recreate setup that I had on my work project in another fresh repo and `mypy` was arguing on the very same line, even though that mypy/qdrant lib versions were the same. You were right, mypy is also arguing on invariant type issue
